src/app/clusters/service-area-server/service-area-delegate.h

```diff
@@ -0,0 +1,384 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "service-area-cluster-objects.h"
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ServiceArea {
+
+class Instance;
+
+// *****************************************************************************
+// cluster constraints
+
+constexpr size_t kMaxNumSupportedLocations = 255;
+constexpr size_t kMaxNumSupportedMaps      = 255;
+constexpr size_t kMaxNumSelectedLocations  = 255;
+constexpr size_t kMaxNumProgressElements   = 255;
+
+constexpr size_t kMaxSizeStatusText = 256;
+
+/**
+ * ServiceArea::Delegate Defines methods for implementing application-specific
+ * logic for the Service Area Cluster.
+ */
+class Delegate
+{
+public:
+    Delegate()          = default;
+    virtual ~Delegate() = default;
+
+    /**
+     * Stop this class objects from being copied.
```

Comment:
Comment about why? what self-referential data structures are contained within this class?

Reply:
I see mInstance however I am unclear about what management of it is done. at a frist glance it seems to be just a raw pointer with set/get method.

Reply:
`mInstance` points to an `Instance` object that has a pointer to this `Delegate` object. The `Delegate` and `Instance` classes are coupled in this way. 

Copying a delegate class might make things confusing as it allows the possibility of a `Delegate` object that points to an `Instance` that does not point back to it.

Reply:
This should be explained in a comment. A `stop this class  object for being copied` just re-states the code when I see a `= delete`. When maintaining this code, I need to answer why this is not allowed and what would happen if I do allow it.

Design-wise, the circular reference is a bit frustrating.

Reply:
Added an explanation in [this commit](https://github.com/project-chip/connectedhomeip/pull/34738/commits/b73083277a6df686bdcc6f071dfc4835f586ea0a).